Preserve sign of NaN operands in fsub{32,64}

Achierius opened the pull request:
#### 5f1ac6454783b0c77f1dc789965e7a097e948771
<pre>
Preserve sign of NaN operands in fsub{32,64}
<a href="https://bugs.webkit.org/show_bug.cgi?id=269509">https://bugs.webkit.org/show_bug.cgi?id=269509</a>
<a href="https://rdar.apple.com/120780768">rdar://120780768</a>

Reviewed by Justin Michaud and Yusuke Suzuki.

Negating a NaN operand for fsub is incorrect, as IEEE defines the result
of an arithmetic operation with a NaN parameter as the NaN itself
(without modification). This is observable when a copysign operation is
done on the result, as copysign itself is exempt from that restriction
and can thus detect the improper negation.
By checking whether the operand is a NaN before negating we avoid this
issue.

* JSTests/wasm/stress/fsub-nan-copysign.js: Added.
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJITImpl::BBQJIT::addF32Sub):
(JSC::Wasm::BBQJITImpl::BBQJIT::addF64Sub):

Canonical link: <a href="https://commits.webkit.org/274910@main">https://commits.webkit.org/274910@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e365a6c142c4808c81f34c9cd0fd6766ce2e6ea0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/40270 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/19282 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/42648 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/42815 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/36359 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/22212 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/16611 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/33442 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40844 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/16239 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/34750 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/14015 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/14066 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/35703 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/44093 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/33723 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/36514 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/36033 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/39777 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/39896 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/15085 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/12360 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/38056 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/16704 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/46905 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9054 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/16753 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/9654 "Passed tests") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/16348 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

justinmichaud commented on JSTests/wasm/stress/fsub-nan-copysign.js:
Can we get a test for the non-const and the const-const case too?

Achierius replied:
Done

justinmichaud reviewed (approved):
r=me

Constellation reviewed (approved):
r=me

Achierius committed:
Preserve sign of NaN operands in fsub{32,64}
https://bugs.webkit.org/show_bug.cgi?id=269509
rdar://120780768

Reviewed by Justin Michaud and Yusuke Suzuki.

Negating a NaN operand for fsub is incorrect, as IEEE defines the result
of an arithmetic operation with a NaN parameter as the NaN itself
(without modification). This is observable when a copysign operation is
done on the result, as copysign itself is exempt from that restriction
and can thus detect the improper negation.
By checking whether the operand is a NaN before negating we avoid this
issue.

* JSTests/wasm/stress/fsub-nan-copysign.js: Added.
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJITImpl::BBQJIT::addF32Sub):
(JSC::Wasm::BBQJITImpl::BBQJIT::addF64Sub):

Canonical link: https://commits.webkit.org/274910@main
